charms/unit/harness.py

```diff
@@ -0,0 +1,122 @@
+#
+# harness.py -- Unit Testing Harness for Layer Charms
+#
+# copyright 2016 Canonical Ltd.
+# TODO: License (Apache or LGPL?)
+#
+
+from glob import glob
+import mock
+import unittest
+
+class HarnessError(Exception): pass
+
+class Harness(unittest.TestCase):
+    '''
+    I am a unit testing harness for Juju Layered charms.
+
+    I automatically mock out any modules referenced in the layers
+    directory that do not actually exist in the layer's source. I
+    maintain an internal table of the mock objects, so that you may
+    add behavior to them if necessary for a test.
+
+    I will also setup a mock for the status handlers in hookenv, so
+    that you may set and confirm status in a test.
+
+    '''
+    def __init__(self, *args, **kwargs):
+        super(Harness, self).__init__(*args, **kwargs)
+        self._to_mock = []
+        self._patchers = []
+        self._local_modules = None
+        self.mocks = {}
+
+    @property
+    def last_status(self):
+        '''
+        Helper for mocked out status list.
+
+        Returns the last status set. Raises a HarnessError if this
+        method is called, but the hookenv mock is not active.
+
+        '''
+        try:
+            return self.statuses[-1]
+        except AttributeError:
+            raise HarnessError(
+                'self.statuses does not exist. Did you initialize a '
+                'harness without passing in mock_hookenv_status=True?')
```

Comment:
@johnsca Since the expected values are tuples, what I've found myself doing is things like:

```
self.assertEqual(self.last_status[0], 'active')
```

... which will raise an AttributeError if this routine returns None.

Maybe the best thing to do is to just seed the status list with `(None, None)` as its first element? That way, you always get a comparison that works. I didn't want people to waste time trying to figure out why their code wasn't setting a status, when the problem was that hookenv isn't actually mocked. You do have to deliberately turn if off, though, so maybe that isn't something that we need to protect people from ...